### PR TITLE
TMS320 and TMS320c64x register profile

### PR DESCRIPTION
**Pull request by Rot127**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Adds the register profile for TMS320 and TMS320c64x.

@wargio  Please take a look at https://github.com/rizinorg/rizin/commit/6509a9c48814d945de3e7118873abbcd18e2ccea

Also I failed to find a source for the TMS320 calling convention. So I just added a generic one to suppress "no argument register given" warning.

**Test plan**

I will open a separated PR to add a test after https://github.com/rizinorg/rizin/pull/2806 is merged.

**Closing issues**

...

### CODE REVIEW

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Review by wargio — CHANGES_REQUESTED** [review]
requires the other PR.

**Comment by XVilka** [comment]
@wargio, which one? It seems to work just fine without anything else.

**Review by wargio — APPROVED** [review]

**Comment by wargio** [comment]
@Rot127 https://www.ti.com/lit/an/spra493/spra493.pdf?ts=1657666725198 for the calling convention

image

**Comment by Rot127** [comment]
> @wargio, which one? It seems to work just fine without anything else.

This PR needs https://github.com/rizinorg/rizin/pull/2806 so all registers are added correctly to the profile.
But https://github.com/rizinorg/rizin/pull/2806 needs this PR to test its comment parsing.
~~Could have just based it on #2806.~~ -> Rebased it onto #2806 and added a test

**Comment by Rot127** [comment]
> @Rot127 https://www.ti.com/lit/an/spra493/spra493.pdf?ts=1657666725198 for the calling convention

Thanks. I found this as well. But since it does not match the register names from [Capstone](https://github.com/capstone-engine/capstone/blob/master/arch/TMS320C64x/TMS320C64xGenRegisterInfo.inc#L13=) or the [ISA](https://www.ti.com/lit/ug/spru732j/spru732j.pdf) I did not used it. I don't have time to dig into TMS320 at the moment, so if you are fine with it, I would leave it like that.

**Review by XVilka — CHANGES_REQUESTED** [review]
Build failure on Windows:
```
  Creating library binrz/rz-asm/rz-asm.lib and object binrz/rz-asm/rz-asm.exp
rz_reg.lib(profile.c.obj) : error LNK2019: unresolved external symbol index referenced in function parse_reg_profile_str
binrz/rz-asm/rz-asm.exe : fatal error LNK1120: 1 unresolved externals
[1701/1921] Linking target binrz/rz-hash/rz-hash.exe
FAILED: binrz/rz-hash/rz-hash.exe 
"link"  /MACHINE:x64 /OUT:binrz/rz-hash/rz-hash.exe binrz/rz-hash/rz-hash.exe.p/rz-hash.c.obj "/nologo" "/release" "/nologo" "/OPT:REF" "librz\main\rz_main.lib" "librz\util\rz_util.lib" "subprojects\sdb\src\libsdb.a" "subprojects\zlib-1.2.12\libz.a" "subprojects\xz-5.2.5\src\liblzma\liblzma.a" "librz\demangler\rz_demangler.lib" "subprojects\libdemangle\libdemangle.a" "librz\magic\rz_magic.lib" "librz\socket\rz_socket.lib" "librz\flag\rz_flag.lib" "librz\cons\rz_cons.lib" "librz\hash\rz_hash.lib" "subprojects\xxhash\libxxhash.a" "librz\crypto\rz_crypto.lib" "subprojects\nettle\libnettle-aes.a" "librz\il\rz_il.lib" "librz\reg\rz_reg.lib" "librz\io\rz_io.lib" "subprojects\rzbochs\librzbochs.a" "subprojects\rzgdb\librzgdb.a" "subprojects\rzwinkd\librzwinkd.a" "subprojects\libzip-1.9.2\liblibzip.a" "subprojects\rzar\librzar.a" "subprojects\rzqnx\librzqnx.a" "subprojects\rzw32dbg_wrap\librzw32dbg_wrap.a" "librz\bp\rz_bp.lib" "librz\syscall\rz_syscall.lib" "librz\parse\rz_parse.lib" "librz\asm\rz_asm.lib" "librz\config\rz_config.lib" "librz\bin\rz_bin.lib" "librz\type\rz_type.lib" "subprojects\tree-sitter-c\libtree-sitter-c.a" "subprojects\tree-sitter-0.20.6\libtree_sitter.a" "subprojects\lz4-1.9.3\liblz4.a" "subprojects\yxml\libyxml.a" "subprojects\libmspack\liblibmspack.a" "subprojects\capstone-bundled\libcapstone.a" "subprojects\rzspp\librzspp.a" "librz\egg\rz_egg.lib" "subprojects\mpc\libmpc.a" "librz\search\rz_search.lib" "librz\analysis\rz_analysis.lib" "librz\diff\rz_diff.lib" "librz\debug\rz_debug.lib" "librz\sign\rz_sign.lib" "librz\core\rz_core.lib" "librz\lang\rz_lang.lib" "subprojects\rizin-shell-parser\libshell_parser.a" "subprojects\libuv-v1.40.0\libuv.a" "ws2_32.lib" "wininet.lib" "psapi.lib" "ws2_32.lib" "wininet.lib" "psapi.lib" "advapi32.lib" "bcrypt.lib" "advapi32.lib" "bcrypt.lib" "ws2_32.lib" "wininet.lib" "psapi.lib" "psapi.lib" "user32.lib" "advapi32.lib" "iphlpapi.lib" "userenv.lib" "ws2_32.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
   Creating library binrz/rz-hash/rz-hash.lib and object binrz/rz-hash/rz-hash.exp
rz_reg.lib(profile.c.obj) : error LNK2019: unresolved external symbol index referenced in function parse_reg_profile_str
binrz/rz-hash/rz-hash.exe : fatal error LNK1120: 1 unresolved externals
ninja: build stopped: subcommand failed.
Command exited with code 1
```
https://ci.appveyor.com/project/rizinorg/rizin/builds/44165964/job/68lgv47je8dtm4ep#L2890

**Comment by wargio** [comment]
image

@Rot127 i found it, page 24 of https://www.ti.com/lit/an/sprab89a/sprab89a.pdf?ts=1657804315231&ref_url=https%253A%252F%252Fwww.ti.com%252F
image

Essentially you want A15 and from B3 to B15

**Comment by Rot127** [comment]
@wargio Thank you very much! Worked on this PR only when tired or in a rush. So thanks for compensating my lack of rigor.

**Comment by Rot127** [comment]
https://github.com/rizinorg/rizin/pull/2809 and https://github.com/rizinorg/rizin/pull/2806 should be merged before this.
Or only this and close the upper ones.

**Review by XVilka — APPROVED** [review]
Approving TMS320 part

**Review by wargio — APPROVED** [review]

**Comment by XVilka** [comment]
@Rot127 please rebase and resolve conflicts

**Comment by Rot127** [comment]
This is embarrassing but I just recognized that I forgot the TMS320c55x registers :see_no_evil: Please let me add them before merge.

**Review by wargio — CHANGES_REQUESTED** [review]
wait for missing regs

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Comment by Rot127** [comment]
They need to be lower case.... Today seems to be one of those days...

**Comment by XVilka** [comment]
```
[XX] db/analysis/tms320 arp TMS320 c55x
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.arch=tms320
e analysis.arch=tms320
e analysis.cpu=c55x
arp
ar
' =
-- stdout
--- expected
+++ actual
@@ -67,6 +67,8 @@
 ctr st3_55 .16 559 0 # Status registers 3
 ctr t0     .16 560 0 # Temporary register 0
 ctr t1     .16 561 0 # Temporary register 1
+ctr t2     .16 562 0 # Temporary register 2
+ctr t3     .16 563 0 # Temporary register 3
 ctr trn0   .16 564 0 # Transition registers 1
 ctr trn1   .16 565 0 # Transition registers 1
 ctr xcdp   .23 566 0 # Extended coefficient data pointer
```

**Comment by XVilka** [comment]
```

[XX] db/asm/tms320_c55x_32 <asm> amov 0x22, ac0
-- <asm> amov 0x22, ac0 <--- 142205
-- disassembly
--- expected
+++ actual
@@ -1,1 +1,2 @@
+WARNING: is_c5000: assertion 'cpu' failed (line 116)
 amov 0x22, ac0
```

File: librz/analysis/p/analysis_tms320.c [hunk 1]
```diff
@@ -112,6 +112,246 @@ static bool tms320_fini(void *user) {
 	return true;
 }
 
+static bool is_c5000(const char *cpu) {
+	rz_return_val_if_fail(cpu, false);
```
Comment (wargio) [inline librz/analysis/p/analysis_tms320.c]: allow this to be null and return false.

**Review by wargio — APPROVED** [review]

**Comment by Rot127** [comment]
```diff
[XX] db/formats/pdb Extract pdb via idbx
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'idpi bins/pdb/basic32.pd_
e log.level=3
idpx bins/pdb/basic32.pd_ .tmp
!rz-hash -a md5 .tmp/basic32.pdb
' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-.tmp/basic32.pdb: 0x00000000-0x00649000 md5: a6c609e6b62cd0b9fbea5400bf370891
-- stderr
--- expected
+++ actual
@@ -1,2 +1,0 @@
-ERROR: The pdb file bins/pdb/basic32.pd_ seems to be compressed, please use idpx command to extract the contents.
-INFO: cab_extract: extracted .tmp/basic32.pdb
```

Doesn't seem to be related.

**Review by XVilka — APPROVED** [review]
It isn't

**Pull request merged**